File: backend/app/api/v1/training.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, status
from fastapi.responses import PlainTextResponse
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import CurrentUser, get_current_user, require_staff
from app.db.session import get_db
from app.models.training import (
    TrainingCohort,
    TrainingEnrollment,
    TrainingHours,
    TrainingProgram,
    TrainingRequirement,
)
from app.schemas.training import (
    CohortCreate,
    CohortProgress,
    CohortRead,
    CohortUpdate,
    EnrollmentCreate,
    EnrollmentRead,
    EnrollmentUpdate,
    ProgramCreate,
    ProgramRead,
    ProgramUpdate,
    RequirementCreate,
    RequirementRead,
    TrainingDashboard,
    TrainingHoursCreate,
    TrainingHoursRead,
)
from app.services import training as training_service
from app.services.training import (
    CohortRepository,
    EnrollmentRepository,
    ProgramRepository,
    TrainingHoursRepository,
    TrainingRequirementRepository,
)

router = APIRouter(prefix="/training", tags=["training"])


@router.post("/hours", response_model=TrainingHoursRead, status_code=status.HTTP_201_CREATED)
async def log_hours(
    data: TrainingHoursCreate,
    current: CurrentUser = Depends(require_staff),
    db: AsyncSession = Depends(get_db),
):
    entry = TrainingHours(**data.model_dump())
    return await training_service.log_hours(db, current.tenant_id, entry)


@router.get("/hours/{trainee_id}", response_model=list[TrainingHoursRead])
async def list_hours(
    trainee_id: uuid.UUID,
    current: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await TrainingHoursRepository(db, current.tenant_id).for_trainee(trainee_id)


@router.get("/dashboard/{trainee_id}", response_model=TrainingDashboard)
async def dashboard(
    trainee_id: uuid.UUID,
    current: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await training_service.dashboard(db, current.tenant_id, trainee_id)


@router.get("/export/{trainee_id}.csv", response_class=PlainTextResponse)
async def export_csv(
    trainee_id: uuid.UUID,
    current: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    csv_data = await training_service.export_csv(db, current.tenant_id, trainee_id)
    return PlainTextResponse(
        csv_data,
        media_type="text/csv",
        headers={
            "Content-Disposition": f"attachment; filename=training_{trainee_id}.csv"
        },
    )


@router.post(
    "/requirements", response_model=RequirementRead, status_code=status.HTTP_201_CREATED
)
async def create_requirement(
    data: RequirementCreate,
    current: CurrentUser = Depends(require_staff),
    db: AsyncSession = Depends(get_db),
):
    repo = TrainingRequirementRepository(db, current.tenant_id)
    req = repo.add(TrainingRequirement(**data.model_dump()))
    await db.commit()
    await db.refresh(req)
    return req


@router.get("/requirements", response_model=list[RequirementRead])
async def list_requirements(
    current: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await TrainingRequirementRepository(db, current.tenant_id).list()


# --- Programs ---
@router.post(
    "/programs", response_model=ProgramRead, status_code=status.HTTP_201_CREATED
)
async def create_program(
    data: ProgramCreate,
    current: CurrentUser = Depends(require_staff),
    db: AsyncSession = Depends(get_db),
):
    repo = ProgramRepository(db, current.tenant_id)
    program = repo.add(TrainingProgram(**data.model_dump()))
    await db.commit()
    await db.refresh(program)
    return program


@router.get("/programs", response_model=list[ProgramRead])
async def list_programs(
    current: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await ProgramRepository(db, current.tenant_id).list()


@router.patch("/programs/{program_id}", response_model=ProgramRead)
async def update_program(
    program_id: uuid.UUID,
    data: ProgramUpdate,
    current: CurrentUser = Depends(require_staff),
    db: AsyncSession = Depends(get_db),
):
    repo = ProgramRepository(db, current.tenant_id)
    program = await repo.get(program_id)
    if program is None:
        raise HTTPException(status_code=404, detail="Program not found")
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(program, field, value)
    await db.commit()
    await db.refresh(program)
    return program


# --- Cohorts ---
@router.post(
    "/cohorts", response_model=CohortRead, status_code=status.HTTP_201_CREATED
)
async def create_cohort(
    data: CohortCreate,
    current: CurrentUser = Depends(require_staff),
    db: AsyncSession = Depends(get_db),
):
    program = await ProgramRepository(db, current.tenant_id).get(data.program_id)
    if program is None:
        raise HTTPException(status_code=404, detail="Program not found")
    cohort = CohortRepository(db, current.tenant_id).add(
        TrainingCohort(**data.model_dump())
    )
    await db.commit()
    await db.refresh(cohort)
    return CohortRead(**_cohort_dict(cohort, enrolled_count=0))


@router.get("/cohorts", response_model=list[CohortRead])
async def list_cohorts(
    program_id: uuid.UUID | None = None,
    current: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    repo = CohortRepository(db, current.tenant_id)
    cohorts = (
        await repo.for_program(program_id)
        if program_id is not None
        else await repo.list()
    )
    result = []
    for c in cohorts:
        count = await training_service._enrolled_count(db, current.tenant_id, c.id)
        result.append(CohortRead(**_cohort_dict(c, enrolled_count=count)))
    return result


@router.patch("/cohorts/{cohort_id}", response_model=CohortRead)
async def update_cohort(
    cohort_id: uuid.UUID,
    data: CohortUpdate,
    current: CurrentUser = Depends(require_staff),
    db: AsyncSession = Depends(get_db),
):
    repo = CohortRepository(db, current.tenant_id)
    cohort = await repo.get(cohort_id)
    if cohort is None:
        raise HTTPException(status_code=404, detail="Cohort not found")
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(cohort, field, value)
    await db.commit()
    await db.refresh(cohort)
    count = await training_service._enrolled_count(db, current.tenant_id, cohort.id)
    return CohortRead(**_cohort_dict(cohort, enrolled_count=count))


@router.get("/cohorts/{cohort_id}/progress", response_model=CohortProgress)
async def cohort_progress(
    cohort_id: uuid.UUID,
    current: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    cohort = await CohortRepository(db, current.tenant_id).get(cohort_id)
    if cohort is None:
        raise HTTPException(status_code=404, detail="Cohort not found")
    return await training_service.cohort_progress(db, current.tenant_id, cohort)


# --- Enrollments ---
@router.post(
    "/cohorts/{cohort_id}/enrollments",
    response_model=EnrollmentRead,
    status_code=status.HTTP_201_CREATED,
)
async def enroll_member(
    cohort_id: uuid.UUID,
    data: EnrollmentCreate,
    current: CurrentUser = Depends(require_staff),
    db: AsyncSession = Depends(get_db),
):
    repo = EnrollmentRepository(db, current.tenant_id)
    cohort = await CohortRepository(db, current.tenant_id).get(cohort_id)
    if cohort is None:
        raise HTTPException(status_code=404, detail="Cohort not found")
    existing = await repo.get_for_member(cohort_id, data.member_id)
    if existing is not None:
        raise HTTPException(
            status_code=409, detail="Member already enrolled in this cohort"
        )
    enrollment = repo.add(
        TrainingEnrollment(
            cohort_id=cohort_id,
            member_id=data.member_id,
            enrolled_on=training_service.default_enrolled_on(data.enrolled_on),
            status=data.status,
        )
    )
    await db.commit()
    await db.refresh(enrollment)
    return enrollment


@router.get("/cohorts/{cohort_id}/enrollments", response_model=list[EnrollmentRead])
async def list_enrollments(
    cohort_id: uuid.UUID,
    current: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    enrollments = await EnrollmentRepository(db, current.tenant_id).for_cohort(
        cohort_id
    )
    names = await training_service._member_names(
        db, current.tenant_id, [e.member_id for e in enrollments]
    )
    return [
        EnrollmentRead(
            id=e.id,
            cohort_id=e.cohort_id,
            member_id=e.member_id,
            enrolled_on=e.enrolled_on,
            status=e.status,
            member_name=names.get(e.member_id),
        )
        for e in enrollments
    ]


@router.patch("/enrollments/{enrollment_id}", response_model=EnrollmentRead)
async def update_enrollment(
    enrollment_id: uuid.UUID,
    data: EnrollmentUpdate,
    current: CurrentUser = Depends(require_staff),
    db: AsyncSession = Depends(get_db),
):
    repo = EnrollmentRepository(db, current.tenant_id)
    enrollment = await repo.get(enrollment_id)
    if enrollment is None:
        raise HTTPException(status_code=404, detail="Enrollment not found")
    enrollment.status = data.status
    await db.commit()
    await db.refresh(enrollment)
    return enrollment


@router.delete(
    "/enrollments/{enrollment_id}", status_code=status.HTTP_204_NO_CONTENT
)
async def remove_enrollment(
    enrollment_id: uuid.UUID,
    current: CurrentUser = Depends(require_staff),
    db: AsyncSession = Depends(get_db),
):
    repo = EnrollmentRepository(db, current.tenant_id)
    enrollment = await repo.get(enrollment_id)
    if enrollment is None:
        raise HTTPException(status_code=404, detail="Enrollment not found")
    await repo.delete(enrollment)
    await db.commit()


def _cohort_dict(cohort: TrainingCohort, *, enrolled_count: int) -> dict:
    return {
        "id": cohort.id,
        "program_id": cohort.program_id,
        "name": cohort.name,
        "start_date": cohort.start_date,
        "end_date": cohort.end_date,
        "status": cohort.status,
        "enrolled_count": enrolled_count,
    }

File: frontend/src/api/training.ts
import { api } from './client'
import type {
  Attendance,
  AttendanceStatus,
  CheckIn,
  CohortProgress,
  CohortStatus,
  EnrollmentStatus,
  MemberPass,
  TrainingCohort,
  TrainingDashboard,
  TrainingEnrollment,
  TrainingHours,
  TrainingProgram,
} from '@/types'

// --- Check-in ---
export async function getMemberPass(memberId: string): Promise<MemberPass> {
  const { data } = await api.get<MemberPass>(`/checkin/pass/${memberId}`)
  return data
}

export async function checkInQr(payload: {
  token: string
  session_id: string
  device_id?: string | null
}): Promise<CheckIn> {
  const { data } = await api.post<CheckIn>('/checkin/qr', payload)
  return data
}

export async function checkInManual(payload: {
  member_id: string
  session_id: string
}): Promise<CheckIn> {
  const { data } = await api.post<CheckIn>('/checkin/manual', payload)
  return data
}

// --- Check-in time window (per-studio config) ---
export interface CheckinWindow {
  checkin_opens_before: number
  checkin_closes_after: number
  checkin_late_threshold: number
}

export async function getCheckinWindow(): Promise<CheckinWindow> {
  const { data } = await api.get<CheckinWindow>('/checkin/window')
  return data
}

export async function updateCheckinWindow(
  payload: Partial<CheckinWindow>,
): Promise<CheckinWindow> {
  const { data } = await api.patch<CheckinWindow>('/checkin/window', payload)
  return data
}

// --- Attendance ---
export async function setAttendance(
  sessionId: string,
  payload: { member_id: string; status: AttendanceStatus },
): Promise<unknown> {
  const { data } = await api.put(`/attendance/session/${sessionId}`, payload)
  return data
}

export async function listPendingAttendance(): Promise<Attendance[]> {
  const { data } = await api.get<Attendance[]>('/attendance/pending')
  return data
}

export async function confirmAttendance(
  sessionId: string,
  memberId: string,
): Promise<Attendance> {
  const { data } = await api.post<Attendance>(
    `/attendance/session/${sessionId}/confirm`,
    { member_id: memberId },
  )
  return data
}

export async function rejectAttendance(
  sessionId: string,
  memberId: string,
): Promise<Attendance> {
  const { data } = await api.post<Attendance>(
    `/attendance/session/${sessionId}/reject`,
    { member_id: memberId },
  )
  return data
}

// --- Training ---
export async function getTrainingDashboard(
  traineeId: string,
): Promise<TrainingDashboard> {
  const { data } = await api.get<TrainingDashboard>(
    `/training/dashboard/${traineeId}`,
  )
  return data
}

export async function listTrainingHours(
  traineeId: string,
): Promise<TrainingHours[]> {
  const { data } = await api.get<TrainingHours[]>(`/training/hours/${traineeId}`)
  return data
}

export async function logTrainingHours(payload: {
  trainee_id: string
  area: string
  hours: number
  entry_date: string
  note?: string | null
}): Promise<TrainingHours> {
  const { data } = await api.post<TrainingHours>('/training/hours', payload)
  return data
}

export function trainingCsvUrl(traineeId: string): string {
  const base = import.meta.env.VITE_API_BASE_URL ?? 'http://localhost:8000/api/v1'
  return `${base}/training/export/${traineeId}.csv`
}

// --- Programs ---
export async function listPrograms(): Promise<TrainingProgram[]> {
  const { data } = await api.get<TrainingProgram[]>('/training/programs')
  return data
}

export async function createProgram(payload: {
  name: string
  description?: string | null
  duration_months?: number
}): Promise<TrainingProgram> {
  const { data } = await api.post<TrainingProgram>('/training/programs', payload)
  return data
}

export async function updateProgram(
  programId: string,
  payload: Partial<{
    name: string
    description: string | null
    duration_months: number
    is_active: boolean
  }>,
): Promise<TrainingProgram> {
  const { data } = await api.patch<TrainingProgram>(
    `/training/programs/${programId}`,
    payload,
  )
  return data
}

// --- Cohorts ---
export async function listCohorts(
  programId?: string,
): Promise<TrainingCohort[]> {
  const { data } = await api.get<TrainingCohort[]>('/training/cohorts', {
    params: programId ? { program_id: programId } : undefined,
  })
  return data
}

export async function createCohort(payload: {
  program_id: string
  name: string
  start_date: string
  end_date?: string | null
  status?: CohortStatus
}): Promise<TrainingCohort> {
  const { data } = await api.post<TrainingCohort>('/training/cohorts', payload)
  return data
}

export async function updateCohort(
  cohortId: string,
  payload: Partial<{
    name: string
    start_date: string
    end_date: string | null
    status: CohortStatus
  }>,
): Promise<TrainingCohort> {
  const { data } = await api.patch<TrainingCohort>(
    `/training/cohorts/${cohortId}`,
    payload,
  )
  return data
}

export async function getCohortProgress(
  cohortId: string,
): Promise<CohortProgress> {
  const { data } = await api.get<CohortProgress>(
    `/training/cohorts/${cohortId}/progress`,
  )
  return data
}

// --- Enrollments ---
export async function listEnrollments(
  cohortId: string,
): Promise<TrainingEnrollment[]> {
  const { data } = await api.get<TrainingEnrollment[]>(
    `/training/cohorts/${cohortId}/enrollments`,
  )
  return data
}

export async function enrollMember(
  cohortId: string,
  payload: { member_id: string; enrolled_on?: string; status?: EnrollmentStatus },
): Promise<TrainingEnrollment> {
  const { data } = await api.post<TrainingEnrollment>(
    `/training/cohorts/${cohortId}/enrollments`,
    payload,
  )
  return data
}

export async function updateEnrollment(
  enrollmentId: string,
  status: EnrollmentStatus,
): Promise<TrainingEnrollment> {
  const { data } = await api.patch<TrainingEnrollment>(
    `/training/enrollments/${enrollmentId}`,
    { status },
  )
  return data
}

export async function removeEnrollment(enrollmentId: string): Promise<void> {
  await api.delete(`/training/enrollments/${enrollmentId}`)
}
